[Enhancement] Add a `super-node` graph prediction example (#1021)
awslabs/graphstorm: examples/graph_prediction/README.md

# Example Code for Graph Prediction using GraphStorm
Graph prediction, such as classification and regression, is a common task in Graph Machine
Learning (GML) across various domains, including life sciences and chemistry. In graph prediction,
the entire graph data is typically organized in a batch of subgraphs format, where each subgraph's
nodes have edges only within the subgraph and no edges connecting to nodes in other subgraphs. GML

> labels are linked to these subgraphs. And then GML model makes prediction on new subgraphs as well.

Reviewer: ```suggestion
labels are linked to these subgraphs. Once trained, GML models can makes predictions on new and unseen subgraphs.
```

A typical operation used in graph prediction is called `Read-out`, e.g., `sum`, `mean`, `max` or
`min`, which aggregates the representations of nodes in a subgraph to form one representation for

> the subgraph. Then outputs of the `Read-out` will be used to perform downstream prediction computation.

Reviewer: ```suggestion
the subgraph. The outputs of the `Read-out` can then be used to make predictions downstream, acting as a single representation of the entire subgraph.
```

The current version of GraphStorm can not directly perform graph prediction. But as GraphStorm

> supports node-level prediction, we can use a method called `super-node` to fulfill the graph 
> prediction requirement. Instead of using the `Read-out` operation, we can add a new node, called 

Reviewer: ```suggestion
supports node-level prediction, we can use a method called `super-node` to perform graph-level 
predictions. Instead of using the `Read-out` operation, we can add a new node, called 
```

**super node**, to each subgraph, and link all original nodes of the subgraph to it, but not 
adding reversed edges. With these inbound edges, representations of all original nodes in a 

> subgraph could be easily aggregated to the **super node**. And then we can use the **super node** 
> as the repsentation of this subgraph to perform graph level prediciton tasks. The `super-node` 

Reviewer: ```suggestion
subgraph could be easily aggregated to the **super node**. We can then use the **super node** 
as the representation of this subgraph to perform graph-level prediction tasks. The `super-node` 
```

method help us to turn a graph prediction task into a node prediction task.

> In this example, we demonstrate how to process the common graph property prediction dataset 

Reviewer: ```suggestion
In this example, we demonstrate how to process a common graph property prediction dataset 
```

(the [OGBG datasets](https://ogb.stanford.edu/docs/graphprop/)) into the `super-node` format 

> graph data. And then we can leverage GraphStorm's graph construciton and model trainig CLIs 
> to perform `super node` prediction for graph predicition tasks.

Reviewer: ```suggestion
graph data. We then leverage GraphStorm's graph construction and model training CLIs 
to perform `super node` prediction for graph prediction tasks.
```

## `Super-node` Graph Data Processing

**Step 1**: Generate super-node format OGBG graph data.
``` bash
python gen_ogbg_supernode.py --ogbg-data-name molhiv \
                             --output-path ./supernode_raw/
```

> **Step 2**: Run GraphStorm graph construction CLI.

Reviewer: CLI refers to the interface, we use the GraphStorm CLI to run commands

```suggestion
**Step 2**: Run GraphStorm graph construction command.
```

``` bash
python -m graphstorm.gconstruct.construct_graph \
        --conf-file ./supernode_raw/config.json \
        --output-dir ./supernode_gs_1p/ \
        --num-parts 1 \
        --graph-name supernode_molhiv
```

## Traing GraphStorm GNN Models for `Super-node` Graphs

Using the `super-node` method, we turn the graph prediction task into a node prediction task. 
Then we can leverage GraphStorm's node prediction CLIs to perform the graph prediction on 

> **super nodes**. Below is the CLI for training an RGCN model on the OGBG data.

Reviewer: ```suggestion
**super nodes**. Below is the command for training an RGCN model on the OGBG data.
```

``` bash
python -m graphstorm.run.gs_node_classification \
          --num-trainers 1 \
          --part-config ./supernode_gs_1p/supernode_molhiv.json \
          --cf supernode_gc.yaml \
          --save-model-path ./supernode_gc_model/
```

Users can find the training configuration file, `supernode_gc.yaml`, located at this folder.

> We can also try out other GraphStorm GNN models with CLIs like the followings.

Reviewer: ```suggestion
We can also try out other GraphStorm GNN models with commands like the followings.
```

In this example, we demonstrate how to process the common graph property prediction dataset (the [OGBG datasets](https://ogb.stanford.edu/docs/graphprop/)) into the `super-node` format graph data. And then we can leverage GraphStorm's graph construciton and model trainig CLIs to perform `super node` prediction for graph predicition tasks.

## `Super-node` Graph Data Processing

**Step 1**: Generate super-node format OGBG graph data.
``` bash
python gen_ogbg_supernode.py --ogbg-data-name molhiv \

>                              --output-path ./supernode_raw/

Reviewer: Seems like the core work that needs to happen is handled in this script, but it's not clear to the user what is actually happening behind the scenes. 

I would recommend pulling out some of the code in this Python script to allow readers to follow along, so they can try doing something similar with their own datasets.

Author: As we discussed. I will leave this information to a Notebook in another PR.
